### PR TITLE
Csgenus

### PR DESCRIPTION
**Pull request by simonbrandhorst**
Genus symbols for ZLat

### CODE REVIEW

File: src/QuadForm/Quad/ZGenus.jl [hunk 1]
```diff
@@ -0,0 +1,2001 @@
...
+    U = change_base_ring(ZZ, U * d *lift(u))
+
+    X = C * A
+    A = B * (A - X'*U*X) * B'
+    return [vcat([s[1]+m0] , s[2:end]) for s in vcat(sym,_p_adic_symbol(A, p, val)) ]
+end
+
+
+@doc Markdown.doc"""
+    _two_adic_symbol(A::MatElem) -> Array{Array{Int64,1},1}
```
Comment (thofma) [inline src/QuadForm/Quad/ZGenus.jl:244]: `val` is not described.

File: src/QuadForm/Quad/ZGenus.jl [hunk 2]
```diff
@@ -0,0 +1,2001 @@
...
+    end
+  end
+  return true
+end
+
+@doc Markdown.doc"""
+    representative(S::ZpGenus) -> MatElem
+
+Return a gram matrix of some representative of this local genus.
+    """
```
Comment (thofma) [inline src/QuadForm/Quad/ZGenus.jl:841]:
```suggestion
"""
```

File: src/QuadForm/Quad/ZGenus.jl [hunk 3]
```diff
@@ -0,0 +1,2001 @@
...
+    r = 4*sum(fmpz[min(sym1[m][1], sym1[i][1]) for i in det_differs])
+    if 0 != mod(l-r, 8)
+      return false
+    end
+  end
+  return true
+end
+
+@doc Markdown.doc"""
+    representative(S::ZpGenus) -> MatElem
```
Comment (thofma) [inline src/QuadForm/Quad/ZGenus.jl:838]: For number field lattices this is called `gram_matrix` and `representative` really gives a `ZLat`.

File: src/QuadForm/Quad/ZGenus.jl [hunk 4]
```diff
@@ -0,0 +1,2001 @@
...
+function quadratic_space(G::ZGenus)
+  dim = dimension(G)
+  det = determinant(G)
+  prime_neg_hasse = [prime(s) for s in G._symbols if hasse_invariant(s)==-1]
+  neg = G._signature_pair[2]
+  qf =_quadratic_form_with_invariants(dim, det, prime_neg_hasse, neg)
+  return quadratic_space(QQ, qf)
+end
+
+rational_representative(G::ZGenus) = quadratic_space(G::ZGenus)
```
Comment (thofma) [inline src/QuadForm/Quad/ZGenus.jl:925]:
```suggestion
rational_representative(G::ZGenus) = quadratic_space(G)
```

File: src/QuadForm/Quad/ZGenus.jl [hunk 5]
```diff
@@ -0,0 +1,2001 @@
...
+            S = (ti1 + gen2[i+1].space())-gen1_square[i].space()
+            if not (ti1<=S || ti2<=S)
+                return false
+        return true
+=#
+@doc Markdown.doc"""
+    determinant(S::ZpGenus) -> fmpz
+
+Return an integer representing the determinant of this genus.
+    """
```
Comment (thofma) [inline src/QuadForm/Quad/ZGenus.jl:1056]:
```suggestion
"""
```

File: src/QuadForm/Quad/ZGenus.jl [hunk 6]
```diff
@@ -0,0 +1,2001 @@
...
+    e = _min_nonsquare(p)
+  end
+  return e*prod([ p^(s[1]*s[2]) for s in S._symbol ])
+end
+
+function det(S::ZpGenus)
+ return determinant(S)
+end
+#=
+=#
```
Comment (thofma) [inline src/QuadForm/Quad/ZGenus.jl:1072]:
```suggestion

```

File: src/QuadForm/Quad/ZGenus.jl [hunk 7]
```diff
@@ -0,0 +1,2001 @@
...
+            # can have negative dimension
+            # even if l = L .... && surely
+            # L is represented by itsself
+            S = (ti1 + gen2[i+1].space())-gen1_square[i].space()
+            if not (ti1<=S || ti2<=S)
+                return false
+        return true
+=#
+@doc Markdown.doc"""
+    determinant(S::ZpGenus) -> fmpz
```
Comment (thofma) [inline src/QuadForm/Quad/ZGenus.jl:1053]: This should just be `det` and there should not be a `determinant`.

File: src/QuadForm/Quad/ZGenus.jl [hunk 8]
```diff
@@ -0,0 +1,2001 @@
...
+ return determinant(S)
+end
+#=
+=#
+@doc Markdown.doc"""
+    dimension(S::ZpGenus) -> fmpz
+
+Return the dimension of this genus.
+"""
+function dimension(S::ZpGenus)
```
Comment (thofma) [inline src/QuadForm/Quad/ZGenus.jl:1078]: Just `dim` and no `dimension`.

File: src/QuadForm/Quad/ZGenus.jl [hunk 9]
```diff
@@ -0,0 +1,2001 @@
...
+
+@doc Markdown.doc"""
+    ZGenus
+
+A collection of local genus symbols (at primes)
+and a signature pair. Together they represent the genus of a
+non-degenerate Zlattice.
+"""
+mutable struct ZGenus
+  _signature_pair::Vector{Int}
```
Comment (thofma) [inline src/QuadForm/Quad/ZGenus.jl:66]:
```suggestion
  _signature_pair::Tuple{Int, Int}
```
For immutable lists of fixed length it is more performant to use tuples. They have basically the same semantics, are created b `(a, b)` instead `[a, b]` and are immutable.

If you don't want to adjust it now, we can put it on the todo list.

File: src/QuadForm/Quad/ZGenus.jl [hunk 10]
```diff
@@ -0,0 +1,2001 @@
...
+
+Let `L` be a lattice with bilinear form `b`.
+The norm of `(L,b)` is defined as the ideal
+generated by `\{b(x,x) | x \in L\}`.
+"""
+function norm(G::ZGenus)
+  return prod([norm(s) for s in G._symbols])
+end
+#=
+=#
```
Comment (thofma) [inline src/QuadForm/Quad/ZGenus.jl:1506]:
```suggestion

```

File: src/QuadForm/Quad/ZGenus.jl [hunk 11]
```diff
@@ -0,0 +1,2001 @@
...
+  if n < 0
+    return bernoulli(1-n)//(n-1)
+  elseif n > 1
+    if (n % 2 == 0)
+      return QQ(-1)^(div(n, 2) + 1) * QQ(2)^(n-1) * bernoulli(n) // factorial(ZZ(n))
+    else
+      raise(error("n must be a critical value (i.e. even > 0 or odd < 0)"))
+    end
+  elseif n == 1
+    raise(error("Here is a pole"))
```
Comment (thofma) [inline src/QuadForm/Quad/ZGenus.jl:1958]:
```suggestion
    error("Here is a pole")
```
There is no raise function.

**Review by thofma — COMMENTED** [review]
Thanks a lot. I left some minor things to clean up.

As a global comment, there is no `raise` function. Writing `error("bla")` is enough.

File: src/QuadForm/Quad/ZGenus.jl [hunk 12]
```diff
@@ -358,8 +363,152 @@ function _basis_complement(B::MatElem)
...
+  for k in 0:m
+    if prime(S1) == 2
+      b = [k, 0, 1, 0, 0]
+    else
+      b = [k, 0, 1]
+    end
+    for sym in [sym1, sym2]
+      try
+        s = sym[k]
+        b[2] += s[2]
```
Comment (thofma) [inline src/QuadForm/Quad/ZGenus.jl:461]: How about `haskey(sym, k)`? The `try catch` has some significant overhead.

File: src/QuadForm/Quad/ZGenus.jl [hunk 13]
```diff
@@ -358,8 +363,152 @@ function _basis_complement(B::MatElem)
...
+function direct_sum(S1::ZpGenus, S2::ZpGenus)
+  if prime(S1) != prime(S2)
+    throw(ValueError("the local genus symbols must be over the same prime"))
+  end
+  sym1 = S1._symbol
+  sym2 = S2._symbol
+  m = max(sym1[end][1], sym2[end][1])
+  sym1 = Dict([[s[1], s] for s in sym1])
+  sym2 = Dict([[s[1], s] for s in sym2])
+  symbol = []
```
Comment (thofma) [inline src/QuadForm/Quad/ZGenus.jl:451]:
```suggestion
  symbol = Vector{Int}[]
```

File: src/QuadForm/Quad/ZGenus.jl [hunk 14]
```diff
@@ -358,8 +363,152 @@ function _basis_complement(B::MatElem)
...
+"""
+function direct_sum(S1::ZpGenus, S2::ZpGenus)
+  if prime(S1) != prime(S2)
+    throw(ValueError("the local genus symbols must be over the same prime"))
+  end
+  sym1 = S1._symbol
+  sym2 = S2._symbol
+  m = max(sym1[end][1], sym2[end][1])
+  sym1 = Dict([[s[1], s] for s in sym1])
+  sym2 = Dict([[s[1], s] for s in sym2])
```
Comment (thofma) [inline src/QuadForm/Quad/ZGenus.jl:450]:
```suggestion
  _sym1 = S1._symbol
  _sym2 = S2._symbol
  m = max(_sym1[end][1], _sym2[end][1])
  sym1 = Dict([[s[1], s] for s in _sym1])
  sym2 = Dict([[s[1], s] for s in _sym2])
```

**Review by thofma — COMMENTED** [review]
Sorry, two more minor things.

File: src/QuadForm/Quad/ZGenus.jl (hunk 14 above)
Comment (simonbrandhorst) [reply src/QuadForm/Quad/ZGenus.jl:450]: type stability?

Comment (thofma) [reply src/QuadForm/Quad/ZGenus.jl]: Yes

**Pull request merged**